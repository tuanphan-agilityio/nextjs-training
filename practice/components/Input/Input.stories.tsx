import React from 'react';
import { Meta, Story } from '@storybook/react';

import Input, { InputProps } from '.';

export default {
  title: 'Input',
  component: Input,
} as Meta;

const Template: Story<InputProps> = (args) => <Input {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Name',
  placeholder: 'Enter your name',
};

export const WithEndDecorator = Template.bind({});
WithEndDecorator.args = {
  className: 'max-w-[488px]',
  label: 'Email',
  placeholder: 'Enter your email',
  endDecorator: <div>@example.com</div>,
};
