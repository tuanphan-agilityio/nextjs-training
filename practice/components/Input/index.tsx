import { FC, InputHTMLAttributes, ReactElement, memo } from 'react';
import clsx from 'clsx';

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  endDecorator?: ReactElement;
  className?: string;
}

const Input: FC<InputProps> = ({
  label = '',
  endDecorator = null,
  className = '',
  ...props
}) => {
  return (
    <>
      {label && (
        <label className='block w-full text-md font-bold text-primary leading-7 mb-1'>
          {label}
        </label>
      )}
      <div className={clsx('flex px-4 py-2 w-full bg-primary', className)}>
        <input
          className='block w-full border-none focus:outline-none'
          {...props}
        />
        {endDecorator && (
          <div className='flex items-center w-6 bg-primary'>{endDecorator}</div>
        )}
      </div>
    </>
  );
};

export type { InputProps };

export default memo(Input);
