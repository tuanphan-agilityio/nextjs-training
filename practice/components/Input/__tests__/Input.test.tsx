import { render, screen, fireEvent } from '@testing-library/react';
import Input, { InputProps } from '..';

describe('Input Component', () => {
  const mockProps: InputProps = {
    label: 'Name',
    placeholder: 'Enter your name',
    endDecorator: <div>End</div>,
  };

  it('renders correctly with a label and end decorator', () => {
    const { container } = render(<Input {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly without a label and end decorator', () => {
    const { container } = render(<Input placeholder='Placeholder Text' />);
    expect(container).toMatchSnapshot();
  });

  it('renders the label', () => {
    render(<Input {...mockProps} />);
    const labelElement = screen.getByText('Name');
    expect(labelElement).toBeInTheDocument();
  });

  it('renders the end decorator', () => {
    render(
      <Input
        placeholder='Enter your text'
        endDecorator={<div>@gmail.com</div>}
      />,
    );
    const endDecoratorElement = screen.getByText('@gmail.com');
    expect(endDecoratorElement).toBeInTheDocument();
  });

  it('handles input and extracts input data', () => {
    const { container } = render(<Input {...mockProps} />);

    const inputElement = screen.getByPlaceholderText(
      'Enter your name',
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: 'NextJs Practice' } });

    const inputValue = inputElement.value;
    expect(inputValue).toBe('NextJs Practice');
    expect(container).toMatchSnapshot();
  });
});
