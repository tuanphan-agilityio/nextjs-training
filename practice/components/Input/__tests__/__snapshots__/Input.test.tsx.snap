// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component handles input and extracts input data 1`] = `
<div>
  <label
    class="block w-full text-md font-bold text-primary leading-7 mb-1"
  >
    Name
  </label>
  <div
    class="flex px-4 py-2 w-full bg-primary"
  >
    <input
      class="block w-full border-none focus:outline-none"
      placeholder="Enter your name"
    />
    <div
      class="flex items-center w-6 bg-primary"
    >
      <div>
        End
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders correctly with a label and end decorator 1`] = `
<div>
  <label
    class="block w-full text-md font-bold text-primary leading-7 mb-1"
  >
    Name
  </label>
  <div
    class="flex px-4 py-2 w-full bg-primary"
  >
    <input
      class="block w-full border-none focus:outline-none"
      placeholder="Enter your name"
    />
    <div
      class="flex items-center w-6 bg-primary"
    >
      <div>
        End
      </div>
    </div>
  </div>
</div>
`;

exports[`Input Component renders correctly without a label and end decorator 1`] = `
<div>
  
  <div
    class="flex px-4 py-2 w-full bg-primary"
  >
    <input
      class="block w-full border-none focus:outline-none"
      placeholder="Placeholder Text"
    />
  </div>
</div>
`;
